### PR TITLE
fix pds_compress usage from other module

### PR DESCRIPTION
**Pull request by sl8vz**

### CODE REVIEW

File: test-feature/pds_compress.py [hunk 1]
```diff
@@ -527,7 +527,8 @@ def main(options):
     elif options.out_format == "tlv":
         # We may apply this encoding to all the other outputs. However, the TLV
         # is the only one which generates binary data.
-        options.output.reconfigure(encoding='charmap', newline='')
+        if __name__ == '__main__':
+            options.output.reconfigure(encoding='charmap', newline='')
```
Comment (sl8vz) [inline test-feature/pds_compress.py:531]: Hi @MarcDorval ,

reconfigure is not avail on StringIO()

File: test-feature/wfx_test_target.py [hunk 2]
```diff
@@ -21,9 +21,10 @@
 pds_env['SEND_PDS_FILE'] = "/sys/kernel/debug/ieee80211/phy0/wfx/send_pds"
 pds_env['PDS_DEFINITION_ROOT'] = ""
 pds_env['PDS_DEFINITION_FILE'] = "definitions.in"
-# If using the upstream wfx driver, set below `pds_env['required_options'] = "--out=tlv"`
 pds_env['required_options'] = []
```
Comment (sl8vz) [inline test-feature/wfx_test_target.py:24]: required_options is also propagated to `wfx_test_dut` and `wfx_test_agent`

File: test-feature/wfx_test_target.py [hunk 3]
```diff
@@ -120,7 +122,13 @@ def _prepare_test_data(self, parameters):
 
         pds_string = "#include \"" + pds_env['PDS_DEFINITION_ROOT'] + pds_env['PDS_DEFINITION_FILE']\
                      + "\"\n\n" + pds_compatibility_text + pds_sections
-        compressed_string = compress_string(pds_string, self.required_options)
+
+        compressed_string = compress_string(pds_string, format=self.pds_output_format)
+
+        # The TLV format has some binary data that needs to be included
+        if self.pds_output_format == "tlv":
+            compressed_string = repr(compressed_string).strip("'")
```
Comment (sl8vz) [inline test-feature/wfx_test_target.py:130]: This is to include the non string data from the `tlv` format which I thought was a string but includes some required null bytes.  
Note that `wfx_test_agent` needs to use `echo -e` do decode the `\0x00` 
That's anyway compatible with other pure strings pds formats.

**Comment by sl8vz** [comment]
@MarcDorval 
With these changes I'm finally able to send a somewhat accepted tlv PDS to the firmware BUT it doesn't seem to correctly behave.  
I added the raw string output which is them echoed using `echo -e`. Note that if the PDS is really malformed the firmware just complains and doesn't start any test. Here it seems the overall PDS format is fine but something is still wrong.  

Simple test:
```
from wfx_test_dut import *
dut = WfxTestDut('Local')
dut.tx_mode('GF_MCS0')
dut.channel(7)
dut.tx_start('continuous')

Local: Configuring a Direct connection
Local: fw_version retrieved from HW (3.16.0)
Local: tree filled for FW3.16.0
Direct agent_reply: 1.1.0
PDO\x00{i:{a:7,b:1,f:3E8,c:{a:0,b:0,c:0,d:44},d:{a:BB8,b:0,c:1,d:E,e:0,f:4},e:{}}}
/apps/wfx_test_agent: line 33: echo: write error: Invalid argument
```
Dmesg:
```
[87648.149865] wfx-spi spi0.0: started firmware 3.16.0 "WF200_ASIC_WFM_(Jenkins)_FW3.16.0" (API: 3.11, keyset: C0, caps: 0x00000002)
[87648.182960] wfx-spi spi0.0: MAC address 0: 84:fd:27:1f:b0:4c
[87648.183187] wfx-spi spi0.0: MAC address 1: 84:fd:27:1f:b0:4d
[87648.183333] ieee80211 phy5: Selected rate control algorithm 'minstrel_ht'
[87675.115514] wfx-spi spi0.0: firmware says: Start TX packet test feature

[87675.115908] wfx-spi spi0.0: PDS:1: corrupted file
[87676.117585] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=31 C

[87677.116828] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=31 C

[87678.113987] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=41 C

[87679.119859] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=41 C

[87680.114565] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[87681.119500] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[87682.116129] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=43 C

[87683.114161] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[87684.119185] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[87685.119280] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[87686.117837] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=43 C

[87687.115345] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=43 C

[87688.114225] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=38 C

[87688.167391] wfx-spi spi0.0: firmware assert 1020
[87688.167449] hif: 00000000: bc 04 e0 04 04 00 00 00 fc 03 00 00 00 00 00 00
[87688.167472] hif: 00000010: 00 00 00 00 00 00 00 00 4a a4 4d 02 00 00 00 00
[87688.167492] hif: 00000020: 00 00 00 00 07 db 5d 02 00 04 00 00 00 00 00 00
[87688.167510] hif: 00000030: ea 8b 62 02 00 1e 06 00 00 00 00 00 ec 8b 62 02
[87688.167527] hif: 00000040: 00 1e 0c 00 00 00 00 00 ee 8b 62 02 00 1e 0d 00
[87688.167546] hif: 00000050: 00 00 00 00 f0 8b 62 02 00 9e 00 00 00 00 00 00
[87688.167564] hif: 00000060: 94 01 78 01 00 00 00 00 00 00 00 00 ec b7 80 01
[87688.167582] hif: 00000070: 00 02 00 00 00 00 00 00 d9 c1 80 01 00 00 00 00
[87688.167598] hif: 00000080: 00 00 00 00 75 d2 87 01 00 02 00 00 00 00 00 00
[87688.167614] hif: 00000090: 58 a4 4d 02 02 16 02 00 07 00 01 00 5a a4 4d 02
[87688.167630] hif: 000000a0: 02 1c 02 00 07 00 01 00 5e a4 4d 02 02 0e 02 00
[87688.167647] hif: 000000b0: 01 00 00 00 cb a6 4d 02 02 1c 00 00 00 00 01 00
[87688.167664] hif: 000000c0: cf a6 4d 02 02 1c 01 00 00 00 01 00 6c a7 4d 02
[87688.167682] hif: 000000d0: 02 0a 02 00 00 00 00 00 0c 86 52 02 02 0c 02 00
[87688.167698] hif: 000000e0: 00 00 00 00 6c 87 52 02 02 0a 02 00 00 00 00 00
[87688.167714] hif: 000000f0: 6f 87 5a 02 02 0c 02 00 00 00 00 00 81 87 5a 02
[87688.167733] hif: 00000100: 02 1c 02 00 00 00 00 00 83 87 5a 02 02 10 02 00
[87688.167750] hif: 00000110: 01 00 00 00 25 88 5a 02 02 12 02 00 07 00 02 00
[87688.167766] hif: 00000120: 2a 88 5a 02 02 14 02 00 07 00 02 00 2c 88 5a 02
[87688.167783] hif: 00000130: 02 16 02 00 07 00 02 00 2f 88 5a 02 02 1c 02 00
[87688.167801] hif: 00000140: 07 00 01 00 33 88 5a 02 02 0e 02 00 02 00 00 00
[87688.167817] hif: 00000150: 96 8a 5a 02 02 1c 00 00 00 00 01 00 99 8a 5a 02
[87688.167835] hif: 00000160: 02 1c 01 00 00 00 01 00 37 8b 5a 02 02 0a 02 00
[87688.167851] hif: 00000170: 00 00 00 00 39 8b 62 02 02 0c 02 00 00 00 00 00
[87688.167869] hif: 00000180: 42 8b 62 02 02 1c 02 00 00 00 00 00 44 8b 62 02
[87688.167885] hif: 00000190: 02 90 02 00 02 00 00 00 9f 77 85 01 02 1c 02 00
[87688.167902] hif: 000001a0: 07 00 01 00 a3 77 85 01 02 0e 02 00 0e 00 00 00
[87688.167918] hif: 000001b0: 4c 7a 85 01 02 1c 00 00 00 00 01 00 4f 7a 85 01
[87688.167936] hif: 000001c0: 02 1c 01 00 00 00 01 00 5c d2 87 01 02 1c 02 00
[87688.167952] hif: 000001d0: 00 00 00 00 5e d2 87 01 02 10 02 00 0e 00 00 00
[87688.167968] hif: 000001e0: 50 a4 4d 02 02 12 02 00 07 00 01 00 55 a4 4d 02
[87688.167984] hif: 000001f0: 02 14 02 00 07 00 01 00 cb ba 80 01 03 1a 00 00
[87688.168002] hif: 00000200: 00 00 00 00 4d c2 80 01 03 18 00 2d 31 01 00 00
[87688.168018] hif: 00000210: 25 bf 80 01 03 1a 00 00 00 00 00 00 08 ec 8e 01
[87688.168034] hif: 00000220: 03 18 00 2d 31 01 00 00 26 d6 87 01 03 1a 00 00
[87688.168051] hif: 00000230: 00 00 00 00 b1 dc 87 01 03 18 00 2d 31 01 00 00
[87688.168069] hif: 00000240: 17 da 87 01 03 1a 00 00 00 00 00 00 49 04 88 01
[87688.168084] hif: 00000250: 03 18 00 2d 31 01 00 00 06 58 9b 01 03 9a 00 00
[87688.168102] hif: 00000260: 00 00 00 00 31 ea fe 00 03 1a 00 00 00 00 00 00
[87688.168118] hif: 00000270: e9 f8 fe 00 03 18 00 2d 31 01 00 00 30 f9 77 01
[87688.168135] hif: 00000280: 03 1a 00 00 00 00 00 00 8c 03 78 01 03 18 00 2d
[87688.168152] hif: 00000290: 31 01 00 00 f9 fe 77 01 03 1a 00 00 00 00 00 00
[87688.168167] hif: 000002a0: 92 75 89 01 03 18 00 2d 31 01 00 00 04 7e 5d 02
[87688.168183] hif: 000002b0: 05 00 1a 36 47 02 00 00 a4 8c 5d 02 05 00 1a 36
[87688.168200] hif: 000002c0: 47 02 00 00 44 9b 5d 02 05 00 1a 36 47 02 00 00
```

**Comment by MarcDorval** [comment]
Sebastien,
Do you have running instances of hostapd and/or wpa_supplicant? RF testing is incompatible with wlan use, so you need to kill both if they are running before starting RF tests..
EDIT: firmware assert 1020 is typical of this situation

**Comment by sl8vz** [comment]
Ah yes that must be it, let me check.

**Comment by sl8vz** [comment]
Marc,
Indeed wpa_supplicant was still running !
So it seems WFX does behave correctly. But it still complains that the pds is corrupted (and that correlates to the sysfs error I'm getting).  
Is it a normal behavior ? 

Dmesg:
```
[89200.665963] wfx-spi spi0.0: PDS:1: corrupted file
[89200.708155] wfx-spi spi0.0: firmware says: End of TX packet test

[89313.542619] wfx-spi spi0.0: firmware says: Start TX packet test feature

[89313.542884] wfx-spi spi0.0: PDS:1: corrupted file
[89314.546943] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=32 C

[89315.542122] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=39 C

[89316.544289] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=39 C

[89317.541193] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=41 C

[89318.546875] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=41 C

[89319.544785] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=43 C

[89320.546521] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=43 C

[89321.543181] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[89322.543100] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=42 C

[89323.544030] wfx-spi spi0.0: firmware says: TX packet test ongoing... die temp=43 C

[89324.165571] wfx-spi spi0.0: firmware says: End of TX packet test

[89324.169356] wfx-spi spi0.0: firmware says: Start TX packet test feature

[89324.169623] wfx-spi spi0.0: PDS:1: corrupted file
[89324.541236] wfx-spi spi0.0: firmware says: End of TX packet test
```
And the associated test:
```
Python 3.11.4 (main, Jul 27 2023, 12:44:16) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from wfx_test_dut import *
>>> dut = WfxTestDut('Local')
Local: Configuring a Direct connection
Local: fw_version retrieved from HW (3.16.0)
Local: tree filled for FW3.16.0
Direct agent_reply: 1.1.0
>>> dut.channel(1)
'TEST_CHANNEL_FREQ  1'
>>> dut.tx_mode('GF_MCS0')
'HT_PARAM  GF     RATE  N_MCS0'
>>> dut.tx_start('continuous')
PDO\x00{i:{a:1,b:1,f:3E8,c:{a:0,b:0,c:0,d:44},d:{a:BB8,b:0,c:1,d:E,e:0,f:4},e:{}}}
/apps/wfx_test_agent: line 33: echo: write error: Invalid argument
'TEST_MODE  tx_packet     NB_FRAME  0'
>>> dut.tx_stop()
PDP\x00{i:{a:1,b:1,f:3E8,c:{a:0,b:0,c:0,d:44},d:{a:BB8,b:0,c:1,d:E,e:64,f:4},e:{}}}
/apps/wfx_test_agent: line 33: echo: write error: Invalid argument
'TEST_MODE  tx_packet     NB_FRAME  100'
```

**Comment by sl8vz** [comment]
Hi @MarcDorval ,

By looking at the driver the error was on the next chunk of the TLV (`PDS:1: corrupted file`) which was actually some extra data I was sending.  
Need to use `echo -ne` instead of `echo -e` in `wfx_test_agent` to avoid adding an extra newline. Seems all good now !

**Comment by MarcDorval** [comment]
Sebastien,

One way to check what you are sending is to enable HIF traces, as in https://github.com/SiliconLabs/wfx-linux-tools/blob/SD5/scripts/trace_hif and https://github.com/SiliconLabs/wfx-linux-tools/blob/SD5/scripts/trace_pipe.
This should have helped you see what you were sending as PDS data.

Are your changes keeping compatibility with the more frequent use case, using our wfx driver instead of the upstream driver?
If yes, I'll merge your changes.

**Comment by sl8vz** [comment]
Marc,
Yes the other formats are left unchanged so I don't expect any regression (even if it would be good to test it)
Don't forget that the `wfx_test_agent` needs to be patched for tlv to work (`echo -ne` instead of `echo` in `write_test_data`)

**Comment by MarcDorval** [comment]
@sl8vz , that's what I had in mind. 

Can we find a way to check in wfx_test_agent that the driver is from upstream, and change the behavior accordingly?
I'm not familiar with the upstream driver, sorry.

Marc

**Comment by sl8vz** [comment]
I think that `echo -ne` can be used for both `tlv` and `pds`. Since `pds` is only text, enabling interpretation of backslash escapes `-e` does nothing and I don't think the newline is needed in `pds`

**Comment by sl8vz** [comment]
Also if you can add a tag that helps to get the sources from Buildroot

**Comment by MarcDorval** [comment]
Can you please clarify what you mean by 'add a tag that helps to get the sources from Buildroot'?
Are there specific tag names for that, or is it just easier than using a commit id?

**Comment by MarcDorval** [comment]
(We don't expect updates to these files, we're in a maintenance mode, so using the HEAD should be ok)

**Comment by sl8vz** [comment]
There is no specific name it's just easier

**Pull request merged**